[SEC-1249] Add SHA256 digest variant for iOS13 + CryptoKit

Simon-F-AMS opened the pull request:
- Add digest variant 2 to SHA256 which uses CryptoKit
- Add additional unit test cases for new variants

I'm not sure about the function names yet so suggestions are welcome 👍

Commit:
- Add digest variant 2 to SHA256 which uses CryptoKit
- Add additional unit test cases for new variants

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -41,4 +41,36 @@ class SHA256Tests: XCTestCase {
   }
+
+  @available(iOS 13, *)
+  func testDigestData_iOS13__shouldProduceExpectedHash() {
```
```suggestion
  func testDigestData_usingCryptoKit__shouldProduceExpectedHash() {
```

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -41,4 +41,36 @@ class SHA256Tests: XCTestCase {
+  }
+
+  @available(iOS 13, *)
+  func testDigestFile_iOS13_whenDataLargeEnoughFor5BufferFills__shouldProduceExpectedHash() {
```
```suggestion
  func testDigestFile_usingCryptoKit_whenDataLargeEnoughFor5BufferFills__shouldProduceExpectedHash() {
```

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -41,4 +41,36 @@ class SHA256Tests: XCTestCase {
+  }
+
+  @available(iOS 13, *)
+  func testDigestFile_iOS13_withCustomBufferSize__shouldProduceExpectedHash() {
```
```suggestion
  func testDigestFile_usingCryptoKit_withCustomBufferSize__shouldProduceExpectedHash() {
```

hngry commented on Krypt/Source/SHA256.swift:
```diff
@@ -21,6 +22,13 @@ public struct SHA256 {
+  /// Calculates SHA256 hash of given data using CryptoKit
+  /// - Parameter data: The data to calculate the hash for
+  @available(iOS 13, *)
+  public static func digestV2(_ data: Data) -> Data? {
```
There's no v2, should be part of the `digest(_ data: Data) -> Data?` using either CommonCrypto or CryptoKit depending on iOS version.

hngry commented on Krypt/Source/SHA256.swift:
```diff
@@ -61,4 +69,34 @@ public struct SHA256 {
+  ///   - url: The url to the file to calculate the SHA256 hash for
+  ///   - withBufferSize: The size of the buffer to use in bytes, defaults to 1024*1024 bytes = 1 MB
+  @available(iOS 13, *)
+  public static func digestV2(file url: URL, withBufferSize: Int = 1024 * 1024) throws -> Data {
```
same

hngry commented on Krypt/Source/SHA256.swift:
```diff
@@ -61,4 +69,34 @@ public struct SHA256 {
+    }
+    var hasher = CryptoKit.SHA256()
+    while stream.hasBytesAvailable {
+      let read = stream.read(buffer, maxLength: withBufferSize)
```
Also this one still haunts me, you changed it from `with bufferSize: Int` to `withBufferSize: Int` while I was suggesting `withBufferSize bufferSize: Int` which gives you the best of both.
```suggestion
      let read = stream.read(buffer, maxLength: bufferSize)
```

Simon-F-AMS replied:
my bad will fix it

Simon-F-AMS committed:
Update Example/Tests/SHA256Tests.swift

Co-Authored-By: marko <marko@hungry.dev>

Simon-F-AMS committed:
Update Example/Tests/SHA256Tests.swift

Co-Authored-By: marko <marko@hungry.dev>

Simon-F-AMS committed:
Update Example/Tests/SHA256Tests.swift

Co-Authored-By: marko <marko@hungry.dev>

Commit:
- Drop seperate v2 functions and move crypto kit sha calculation to existing functions wrapped in if  #available(iOS 13, *)
- remove obsolete test functions

Commit:
- Add Conditional CryptoKit import

Commit:
- Adjust podspec so weak framework requirement to CryptoKit is added as an other linker flag
- run pod install to verify iOS < 13 aswell as >= 13 tests run without crashing locally
-

markotect reviewed (approved):
Nice one!
